src/dtree.ml

```diff
@@ -0,0 +1,499 @@
+    let rec loop : rule list -> bool = function
+      | []      -> false
+      | r :: rs -> if is_treecons (fst r.lhs.(k)) then true else loop rs in
+    loop clauses
+
+  (** [discard_cons_free m] returns the list of indexes of columns containing
+      terms that can be matched against (discard constructor-free columns). *)
+  let discard_cons_free : t -> int array = fun m ->
+    let ncols = List.extremum (>)
+      (List.map (fun { lhs ; _ } -> Array.length lhs) m.clauses) in
+    let switchable = List.init ncols (can_switch_on m) in
+    let switchable2ind i e = if e then Some(i) else None in
+    switchable |> List.filteri_map switchable2ind |> Array.of_list
+
+  (** [get_cons l] extracts, sorts and uniqify terms that are tree
+      constructors in [l].  The actual tree constructor (of type
+      {!type:treecons}) is returned along the original term. *)
+  let get_cons : term list -> (treecons * term) list = fun telst ->
+    let keep_treecons e =
+      if is_treecons e then Some(treecons_of_term e, e) else None in
+    let tc_fst_cmp (tca, _) (tcb, _) = tc_compare tca tcb in
+    telst |> List.filter_map keep_treecons |> List.sort_uniq tc_fst_cmp
+
+  (** [contains_abst l] returns whether list of terms [l] contains an
+      abstraction. *)
+  let rec contains_abst : term list -> bool = function
+    | [] -> false
+    | Abst(_, _) :: _  -> true
+    | _          :: xs -> contains_abst xs
+
+  (** [contains_var l] returns whether list of terms [l] contains a pattern
+      variable with a slot assigned. *)
+  let rec contains_var : term list -> bool = function
+    | []                        -> false
+    | Patt(Some(_), _, _) :: _  -> true
+    | _                   :: xs -> contains_var xs
+
+  (** [in_rhs] returns whether a list of term contains a term needed in the
+      right hand side of a rule. *)
+  let rec in_rhs : term list -> bool = function
+    | []                       -> false
+    | Patt(Some(_), _, _) :: _ -> true
+    | _ :: xs                  -> in_rhs xs
+
+  (** [enrich_env_builder c v r] adds an entry into the environment builder of
+      [r] based on the nature of term at column [c] and with [v] variables met
+      so far. *)
+  let enrich_env_builder : int -> int -> rule -> rule = fun ci var_met r ->
+    let t, _ = r.lhs.(ci) in
+    match fst (get_args t) with
+    | Patt(Some(i), _, _) ->
+        { r with env_builder = (var_met, i) :: r.env_builder}
+    | _                   -> r
+
+(** [spec_filter p e] returns whether a line been inspected on element [e]
+    (from a pattern matrix) must be kept when specializing the matrix on
+    pattern [p].
+    @raise Invalid_argument when the element [e] is ill formed. *)
+  let spec_filter : term -> term -> bool = fun pat hd ->
+    let h, args = get_args hd in
+    let ph, pargs = get_args pat in
+    match ph, h with
+    | Symb(_, _), Symb(_, _)
+    | Vari(_)   , Vari(_)       -> List.same_length args pargs && eq ph h
```

Comment:
What is eq here? Thanks to qualify it.

Reply:
Case Vari: Do we really allow matching on bound variables? Do we have any example?

Reply:
Not really indeed, I'll check whether it is relevant